From 7eedb9f28018538968a2f72fda7822566a158b22 Mon Sep 17 00:00:00 2001
From: Johannes Englisch <cyberjoe0815@hotmail.com>
Date: Sat, 1 Dec 2018 23:56:48 +0100
Subject: [PATCH] Add `togglefullscreen` command.

Problem: dwm lacks a 'full screen' command.

Solution: Write a command around dwm's `setfullscreen` function.
---
 config.def.h | 69 ++++++++++++++++++++++++++--------------------------
 dwm.c        |  9 +++++++
 2 files changed, 44 insertions(+), 34 deletions(-)

diff --git a/config.def.h b/config.def.h
index 1c0b587..893c132 100644
--- a/config.def.h
+++ b/config.def.h
@@ -60,40 +60,41 @@ static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont,
 static const char *termcmd[]  = { "st", NULL };
 
 static Key keys[] = {
-	/* modifier                     key        function        argument */
-	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
-	{ MODKEY|ShiftMask,             XK_Return, spawn,          {.v = termcmd } },
-	{ MODKEY,                       XK_b,      togglebar,      {0} },
-	{ MODKEY,                       XK_j,      focusstack,     {.i = +1 } },
-	{ MODKEY,                       XK_k,      focusstack,     {.i = -1 } },
-	{ MODKEY,                       XK_i,      incnmaster,     {.i = +1 } },
-	{ MODKEY,                       XK_d,      incnmaster,     {.i = -1 } },
-	{ MODKEY,                       XK_h,      setmfact,       {.f = -0.05} },
-	{ MODKEY,                       XK_l,      setmfact,       {.f = +0.05} },
-	{ MODKEY,                       XK_Return, zoom,           {0} },
-	{ MODKEY,                       XK_Tab,    view,           {0} },
-	{ MODKEY|ShiftMask,             XK_c,      killclient,     {0} },
-	{ MODKEY,                       XK_t,      setlayout,      {.v = &layouts[0]} },
-	{ MODKEY,                       XK_f,      setlayout,      {.v = &layouts[1]} },
-	{ MODKEY,                       XK_m,      setlayout,      {.v = &layouts[2]} },
-	{ MODKEY,                       XK_space,  setlayout,      {0} },
-	{ MODKEY|ShiftMask,             XK_space,  togglefloating, {0} },
-	{ MODKEY,                       XK_0,      view,           {.ui = ~0 } },
-	{ MODKEY|ShiftMask,             XK_0,      tag,            {.ui = ~0 } },
-	{ MODKEY,                       XK_comma,  focusmon,       {.i = -1 } },
-	{ MODKEY,                       XK_period, focusmon,       {.i = +1 } },
-	{ MODKEY|ShiftMask,             XK_comma,  tagmon,         {.i = -1 } },
-	{ MODKEY|ShiftMask,             XK_period, tagmon,         {.i = +1 } },
-	TAGKEYS(                        XK_1,                      0)
-	TAGKEYS(                        XK_2,                      1)
-	TAGKEYS(                        XK_3,                      2)
-	TAGKEYS(                        XK_4,                      3)
-	TAGKEYS(                        XK_5,                      4)
-	TAGKEYS(                        XK_6,                      5)
-	TAGKEYS(                        XK_7,                      6)
-	TAGKEYS(                        XK_8,                      7)
-	TAGKEYS(                        XK_9,                      8)
-	{ MODKEY|ShiftMask,             XK_q,      quit,           {0} },
+	/* modifier                     key        function          argument */
+	{ MODKEY,                       XK_p,      spawn,            {.v = dmenucmd } },
+	{ MODKEY|ShiftMask,             XK_Return, spawn,            {.v = termcmd } },
+	{ MODKEY,                       XK_b,      togglebar,        {0} },
+	{ MODKEY,                       XK_j,      focusstack,       {.i = +1 } },
+	{ MODKEY,                       XK_k,      focusstack,       {.i = -1 } },
+	{ MODKEY,                       XK_i,      incnmaster,       {.i = +1 } },
+	{ MODKEY,                       XK_d,      incnmaster,       {.i = -1 } },
+	{ MODKEY,                       XK_h,      setmfact,         {.f = -0.05} },
+	{ MODKEY,                       XK_l,      setmfact,         {.f = +0.05} },
+	{ MODKEY,                       XK_Return, zoom,             {0} },
+	{ MODKEY,                       XK_Tab,    view,             {0} },
+	{ MODKEY|ShiftMask,             XK_c,      killclient,       {0} },
+	{ MODKEY,                       XK_t,      setlayout,        {.v = &layouts[0]} },
+	{ MODKEY,                       XK_f,      setlayout,        {.v = &layouts[1]} },
+	{ MODKEY,                       XK_m,      setlayout,        {.v = &layouts[2]} },
+	{ MODKEY,                       XK_space,  setlayout,        {0} },
+	{ MODKEY|ShiftMask,             XK_space,  togglefloating,   {0} },
+	{ MODKEY|ShiftMask,             XK_f,      togglefullscreen, {0} },
+	{ MODKEY,                       XK_0,      view,             {.ui = ~0 } },
+	{ MODKEY|ShiftMask,             XK_0,      tag,              {.ui = ~0 } },
+	{ MODKEY,                       XK_comma,  focusmon,         {.i = -1 } },
+	{ MODKEY,                       XK_period, focusmon,         {.i = +1 } },
+	{ MODKEY|ShiftMask,             XK_comma,  tagmon,           {.i = -1 } },
+	{ MODKEY|ShiftMask,             XK_period, tagmon,           {.i = +1 } },
+	TAGKEYS(                        XK_1,                        0)
+	TAGKEYS(                        XK_2,                        1)
+	TAGKEYS(                        XK_3,                        2)
+	TAGKEYS(                        XK_4,                        3)
+	TAGKEYS(                        XK_5,                        4)
+	TAGKEYS(                        XK_6,                        5)
+	TAGKEYS(                        XK_7,                        6)
+	TAGKEYS(                        XK_8,                        7)
+	TAGKEYS(                        XK_9,                        8)
+	{ MODKEY|ShiftMask,             XK_q,      quit,             {0} },
 };
 
 /* button definitions */
diff --git a/dwm.c b/dwm.c
index 4465af1..754076a 100644
--- a/dwm.c
+++ b/dwm.c
@@ -211,6 +211,7 @@ static void tagmon(const Arg *arg);
 static void tile(Monitor *);
 static void togglebar(const Arg *arg);
 static void togglefloating(const Arg *arg);
+static void togglefullscreen(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
 static void unfocus(Client *c, int setfocus);
@@ -1719,6 +1720,14 @@ togglefloating(const Arg *arg)
 	arrange(selmon);
 }
 
+void
+togglefullscreen(const Arg *arg)
+{
+	if (!selmon->sel)
+		return;
+	setfullscreen(selmon->sel, !selmon->sel->isfullscreen);
+}
+
 void
 toggletag(const Arg *arg)
 {
-- 
2.19.2

